homeassistant/components/dlna_dms/config_flow.py

```diff
@@ -0,0 +1,171 @@
+"""Config flow for DLNA DMS."""
+from __future__ import annotations
+
+import logging
+from pprint import pformat
+from typing import Any, cast
+from urllib.parse import urlparse
+
+from async_upnp_client.profiles.dlna import DmsDevice
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.components import ssdp
+from homeassistant.const import CONF_DEVICE_ID, CONF_HOST, CONF_URL
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.exceptions import IntegrationError
+
+from .const import DEFAULT_NAME, DOMAIN
+
+LOGGER = logging.getLogger(__name__)
+
+
+class ConnectError(IntegrationError):
+    """Error occurred when trying to connect to a device."""
+
+
+class DlnaDmsFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a DLNA DMS config flow.
+
+    The Unique Service Name (USN) of the DMS device is used as the unique_id for
+    config entries and for entities. This USN may differ from the root USN if
+    the DMS is an embedded device.
+    """
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize flow."""
+        self._discoveries: dict[str, ssdp.SsdpServiceInfo] = {}
+        self._location: str | None = None
+        self._usn: str | None = None
+        self._name: str | None = None
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle a flow initialized by the user by listing unconfigured devices."""
+        LOGGER.debug("async_step_user: user_input: %s", user_input)
+
+        if user_input is not None and (host := user_input.get(CONF_HOST)):
+            # User has chosen a device
+            discovery = self._discoveries[host]
+            await self._async_set_info_from_discovery(discovery)
+            return self._create_entry()
+
+        if not (discoveries := await self._async_get_discoveries()):
+            # Nothing found, abort configuration
+            return self.async_abort(reason="no_devices_found")
+
+        self._discoveries = {
+            discovery.upnp.get(ssdp.ATTR_UPNP_FRIENDLY_NAME)
+            or cast(str, urlparse(discovery.ssdp_location).hostname): discovery
+            for discovery in discoveries
+        }
+
+        data_schema = vol.Schema(
+            {vol.Optional(CONF_HOST): vol.In(self._discoveries.keys())}
+        )
+        return self.async_show_form(step_id="user", data_schema=data_schema)
+
+    async def async_step_ssdp(self, discovery_info: ssdp.SsdpServiceInfo) -> FlowResult:
+        """Handle a flow initialized by SSDP discovery."""
+        LOGGER.debug("async_step_ssdp: discovery_info %s", pformat(discovery_info))
+
+        await self._async_set_info_from_discovery(discovery_info)
```

Comment:
Should you move this after where we know the device can be used ?

Reply:
Probably not; `_async_set_info_from_discovery` does some early parsing of the `discovery_info` and does the checks for unique IDs. I've renamed the method to reflect its actual role and added a comment, so it's hopefully less confusing now.